core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -2319,6 +2325,18 @@ public Expression implement(
           return Expressions.call(SqlFunctions.class, backupMethodName,
               expressions);
         }
+        // When check (not) equals on two primitive boxing class (e.g., Long x, Long y),
+        // if "x" or "y" is not a constant null expression, we should fall back to
+        // call "SqlFunctions.eq(x, y)" and "SqlFunctions.ne(x, y)", rather than "x == y"
+        final Primitive boxPrimitive0 = Primitive.ofBox(type0);
+        final Primitive boxPrimitive1 = Primitive.ofBox(type1);
+        if (EQUALS_OPERATORS.contains(op)
+            && boxPrimitive0 != null && boxPrimitive1 != null
+            && !isConstantNullExpression(expressions.get(0))
+            && !isConstantNullExpression(expressions.get(1))) {
```

Comment:
How about if one operand is `Long` and another is `Integer` but with the same value ?

Reply:
In `RexImpTable`, when defining `BinaryImplementor`, there is a `harmonize` function called to ensure that operands have identical type.

Reply:
Well, the `defineBinary` indeed is with `harmonize ` as true, thanks for the explanation.